nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsDefaultDecrypter.java

```diff
@@ -0,0 +1,127 @@
+    }
+
+    public KmsDefaultDecrypter(@NonNull final AWSKMS kms,
+                               @NonNull final String keyId) {
+        super(kms, keyId);
+    }
+
+    public KmsDefaultDecrypter(@NonNull final AWSKMS kms,
+                               @NonNull final String keyId,
+                               @NonNull final Set<String> defCritHeaders) {
+        this(kms, keyId);
+        critPolicy.setDeferredCriticalHeaderParams(defCritHeaders);
+    }
+
+    public KmsDefaultDecrypter(@NonNull final AWSKMS kms,
+                               @NonNull final String keyId,
+                               @NonNull final Map<String, String> encryptionContext,
+                               @NonNull final Set<String> defCritHeaders) {
+        this(kms, keyId, encryptionContext);
+        critPolicy.setDeferredCriticalHeaderParams(defCritHeaders);
+    }
+
+    @Override
+    public Set<String> getProcessedCriticalHeaderParams() {
+        return critPolicy.getProcessedCriticalHeaderParams();
+    }
+
+    @Override
+    public Set<String> getDeferredCriticalHeaderParams() {
+        return critPolicy.getDeferredCriticalHeaderParams();
+    }
+
+    @Override
+    public byte[] decrypt(
+            @NonNull final JWEHeader header,
+            @NonNull final Base64URL encryptedKey,
+            @NonNull final Base64URL iv,
+            @NonNull final Base64URL cipherText,
+            @NonNull final Base64URL authTag)
+            throws JOSEException {
+
+        validateJWEHeader(header);
+        critPolicy.ensureHeaderPasses(header);
+
+        final DecryptResult cekDecryptResult =
+                decryptCek(getKeyId(), getEncryptionContext(), header.getAlgorithm(), encryptedKey);
+        final SecretKey cek =
+                new SecretKeySpec(cekDecryptResult.getPlaintext().array(), header.getAlgorithm().toString());
+        return ContentCryptoProvider.decrypt(header, encryptedKey, iv, cipherText, authTag, cek, getJCAContext());
+    }
+
+    private DecryptResult decryptCek(String keyId, Map<String, String> encryptionContext, JWEAlgorithm alg,
+                                     Base64URL encryptedKey)
+            throws JOSEException {
+        try {
+            return getKms().decrypt(new DecryptRequest()
+                    .withEncryptionContext(encryptionContext)
+                    .withKeyId(keyId)
+                    .withEncryptionAlgorithm(alg.getName())
+                    .withCiphertextBlob(ByteBuffer.wrap(encryptedKey.decode())));
+        } catch (NotFoundException | DisabledException | InvalidKeyUsageException | KeyUnavailableException
+                 | KMSInvalidStateException e) {
+            throw new RemoteKeySourceException("An exception was thrown from KMS due to invalid key.", e);
+        } catch (DependencyTimeoutException | InvalidGrantTokenException | KMSInternalException e) {
+            throw new TemporaryJOSEException("A temporary error was thrown from KMS.", e);
+        }
```

Comment:
This code looks mostly same as the decryption code in the symmetric key decryption class. Can we common out the code?

Reply:
Modified and moved common code to decrypter util